### PR TITLE
Do not include esformatter

### PR DESCRIPTION
**Pull request by nicolai86**
Bundling esformatter leads to all sorts of problems when the user wants to use different plugins. Instead, assume the user has esformatter installed globally via `npm install -g esformatter`.

This means esformatter can be updated without updating this plugin, and it also means the user can configure esformatter more freely using `.esformatter` file.

### CODE REVIEW

**Comment by piuccio** [comment]
I like the idea, I need more time to review it. I see that it creates a temp file. Does it work when you have multiple selections?

Thanks for the PR

**Comment by nicolai86** [comment]
@piuccio I played around with it - everything seems to work. Multiple regions, single regions, entire file.

File: EsFormatter.py [hunk 1]
```diff
@@ -60,8 +99,6 @@ def run(self, edit, save=False, ignoreSelection=False):
 
     def replaceFile(self, thread, save=False):
         '''Replace the entire file content with the formatted text.'''
-        if thread.code == thread.result.encode('utf-8'):
```
Comment (piuccio) [inline EsFormatter.py]: I think these two lines should stay, they are a fix for #43

File: EsFormatter.py [hunk 2]
```diff
@@ -97,8 +134,6 @@ def replaceSelections(self, threads, lastError):
             offset = 0
             regions = []
             for thread in sorted(threads, key=lambda t: t.region.begin()):
-                if thread.code == thread.result.encode('utf-8'):
```
Comment (piuccio) [inline EsFormatter.py]: same as above

File: EsFormatter.py [hunk 3]
```diff
@@ -139,18 +174,22 @@ def handle_threads(self, threads, callback, process=False, lastError=None):
...
-        self.cmd = getNodeCommand(exec_path, path, options)
         self.result = None
         threading.Thread.__init__(self)
 
     def run(self):
         try:
-            process = subprocess.Popen(self.cmd, bufsize=160*len(self.code), stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, startupinfo=getStartupInfo())
+            f = tempfile.NamedTemporaryFile(delete=False)
+            f.write(self.code)
+            f.close()
```
Comment (piuccio) [inline EsFormatter.py]: probably the file should be deleted after calling Popen otherwise we keep creating new files

**Comment by piuccio** [comment]
The code throws an exception on Windows + ST3, it might be related to path and backward slash, I don't have a fix yet, hopefully I have more time later

**Comment by nicolai86** [comment]
@piuccio thanks for the first feedback. I've updated the code accordingly. Sadly I only have OS X + ST3 at hand, so I can only guess at why it crashes on windows, same as you.

File: EsFormatter.py [hunk 4]
```diff
@@ -139,18 +178,22 @@ def handle_threads(self, threads, callback, process=False, lastError=None):
...
         threading.Thread.__init__(self)
 
     def run(self):
+        f = tempfile.NamedTemporaryFile(delete=False)
         try:
-            process = subprocess.Popen(self.cmd, bufsize=160*len(self.code), stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, startupinfo=getStartupInfo())
+            f.write(self.code)
+            f.close()
+
+            process = subprocess.Popen(["esformatter", f.name], bufsize=160*len(self.code), stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, startupinfo=getStartupInfo())
```
Comment (piuccio) [inline EsFormatter.py]: I've got it working on windows with this line

```
process = subprocess.Popen(["esformatter", f.name], bufsize=160*len(self.code), stdout=subprocess.PIPE, stderr=subprocess.PIPE, startupinfo=getStartupInfo(), shell=True)
```

can you check if it works on mac?

File: EsFormatter.py [hunk 5]
```diff
@@ -139,18 +178,22 @@ def handle_threads(self, threads, callback, process=False, lastError=None):
...
             if ST2:
                 stdout, stderr = process.communicate(self.code)
```
Comment (piuccio) [inline EsFormatter.py]: and we don't need to `communicate(self.code)` anymore, `communicate()` is enough, the same couple lines below

File: EsFormatter.py (hunk 4 above)
Comment (nicolai86) [inline EsFormatter.py]: it works with shell=True on mac, too.

**Comment by nicolai86** [comment]
@piuccio I had to remove the tempfile usage, it turns out the esformatter takes input from STDIN, too. It turns out the tempfile wasn't used, instead the code passed to communicate was being used.

**Comment by piuccio** [comment]
I finally had some time to test it on Windows, works fine. Thanks for the PR

**Comment by nicolai86** [comment]
:+1:

**Comment by holm** [comment]
Would be lovely with a release of this!

**Comment by piuccio** [comment]
Yeah, I had an issue with my mac + sublime 3 so I've decided to postpone until I can investigate better. 
If you've got the same setup you can try installing from master branch and let me know if it works fine for you. 
This multi platform is a nightmare

**Comment by holm** [comment]
I just tried it. I had to use https://github.com/int3h/SublimeFixMacPath to get it to pick up esformatter automatically. Unfortunately it no longer seems that it picks up the project roots .esformatter file, so the formatting is done using the defaults.

Not sure if I am doing something wrong, or the path fixer is the cause of it, but I never had problems with it before.

**Pull request merged**